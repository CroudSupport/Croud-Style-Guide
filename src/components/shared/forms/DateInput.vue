<template>
    <div class="ui input">
        <input type="date" v-model="date" :min="minDate" :max="maxDate"/>
    </div>
</template>

<script>
    import moment from 'moment'

    /**
     * Date input
     *
     * @example ./croud-date-input.md
     */
    export default {
        name: 'croud-date-input',

        props: {
            /**
            * Alias for v-model, value of the date input as a moment object or an iso string
            */
            value: {
                type: [Object, String],
                required: true,
            },

            /**
             * Minimum date as a moment object or an iso string
             */
            min: {
                type: [Object, String],
            },

            /**
             * Maximum date as a moment object or an iso string
             */
            max: {
                type: [Object, String],
            },
        },

        data() {
            return {
                format: 'YYYY-MM-DD',
            }
        },

        computed: {
            date: {
                get() {
                    return moment(this.value).format(this.format)
                },
                set(val) {
                    const date = moment(val, this.format, true)

                    if (!date.isValid()) {
                        this.$emit('invalid-date', date)
                    } else if ((this.max && date.isAfter(moment(this.max), 'd'))
                                || (this.min && date.isBefore(moment(this.min), 'd'))) {
                        this.$emit('validation-error', date)
                    }

                    this.$emit('input', date)
                },
            },

            minDate() {
                return this.min ? moment(this.min).format(this.format) : null
            },

            maxDate() {
                return this.max ? moment(this.max).format(this.format) : null
            },
        },
    }
</script>
