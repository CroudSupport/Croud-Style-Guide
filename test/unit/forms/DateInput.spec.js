import Vue from 'vue'
import moment from 'moment'
import DateInput from '../../../src/components/shared/forms/DateInput'

const Constructor = Vue.extend(DateInput)

const vm = new Constructor({
    propsData: {
        value: moment(),
    },
}).$mount()

const input = jest.fn()
const invalid = jest.fn()
const validation = jest.fn()
vm.$on('input', input)
vm.$on('invalid-date', invalid)
vm.$on('validation-error', validation)

describe('Date Input', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })

    describe('events', () => {
        beforeEach(() => {
            input.mockReset()
            invalid.mockReset()
            validation.mockReset()
        })
        it('should fire input event when date changes', (done) => {
            vm.date = moment().add(1, 'd')
            vm.$nextTick(() => {
                expect(input).toBeCalled()
                done()
            })
        })

        it('should fire invalid-date event when date is cleared', (done) => {
            vm.date = ''
            vm.$nextTick(() => {
                expect(invalid).toBeCalled()
                done()
            })
        })

        describe('validation', () => {
            it('should fire validation-error event if sooner than min date', (done) => {
                vm.min = moment().add(1, 'd')
                vm.max = null
                vm.date = moment()
                vm.$nextTick(() => {
                    expect(validation).toBeCalled()
                    done()
                })
            })

            it('should fire validation-error event if later than max date', (done) => {
                vm.min = null
                vm.max = moment().subtract(1, 'd')
                vm.date = moment()

                vm.$nextTick(() => {
                    expect(validation).toBeCalled()
                    done()
                })
            })

            it('should not fire validation-error if between min and max', (done) => {
                vm.min = moment().subtract(1, 'd')
                vm.max = moment().add(1, 'd')
                vm.date = moment()

                vm.$nextTick(() => {
                    expect(validation).not.toBeCalled()
                    done()
                })
            })
        })
    })
})
