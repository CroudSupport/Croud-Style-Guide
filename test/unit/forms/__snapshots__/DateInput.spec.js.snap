// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Input should match the snapshot 1`] = `
<div
  class="ui input"
>
  <input
    type="date"
  />
</div>
`;
